src/System.Management.Automation/engine/runtime/Binding/Binders.cs

```diff
@@ -2029,36 +2029,34 @@ private static object NullRule(CallSite site, object obj)
 
         internal static bool IsValueTypeMutable(Type type)
         {
-            if (type.IsPrimitive || type.IsEnum)
+            // First, check compiler-defined attributes. This could be the builtin type, or a shoe-in by the compiler
+            // if the library targets netstandard2.0, where the compiler can't be sure the attribute will be provided
+            // by the runtime, and defines its own attribute of the same name during compilation.
+            if (type.IsDefined(typeof(System.Runtime.CompilerServices.IsReadOnlyAttribute), inherit: false))
             {
                 return false;
             }
 
-            // If there are any fields, the type is mutable.
-            if (type.GetFields(BindingFlags.Public | BindingFlags.Instance).Any())
+            foreach (var attribute in type.GetCustomAttributes(inherit: false))
             {
-                return true;
+                if (attribute.GetType().FullName.Equals(
+                    "System.Runtime.CompilerServices.IsReadOnlyAttribute",
+                    StringComparison.Ordinal))
+                {
+                    return false;
```

Comment:
The custom attribute targets old runtimes.
It is for edge scenarios to target old runtimes with modern compiler.
It will lose value over time.
So the addition could only have value for Windows PowerShell but not for PowerShell Core and we can remove it.

Reply:
Binary modules targeting `netstandard2.0` will still generate the custom attribute as the compiler can't be sure the attribute will be present when the DLL is loaded.

Reply:
>  Doing so effectively means we no longer support cross-compatibility with WinPS modules at all

No :-) In this case, we do not need to look back. What is current status of the code? 
- no security issues
- no performance issues
- no functional issues (do you know real broken scenarios? I think they don’t exist at all because it’s all for performance)

So only our intention is _to benefit from modern runtime to reduce false positives_.
I think "world wide plan" is to move community to .Net 6.0 LTS. That means Netstandard2.1. In any case, the old code will _decrease_ over time, and if we did not address (the performance) it earlier, then it does not make sense to address it to the future paying with performance due to new code without an urgent need.

Reply:
Not sure I follow you through to the conclusion there. Are you saying we should consciously break the compatibility with netstandard2.0 here and now, or are you saying we should leave behind the old, ineffective code as a fallback?

If you're saying it's not worth changing at all... I don't really agree. PS has to handle mutable from immutable structs a bit differently in a few cases that I see in the code. Being able to yield less false positives from this method should (in theory) allow the engine to take the more efficient code path more often -- though, certainly, the cases where we need to be working with structs in PS directly aren't that common.

Reply:
> Are you saying we should consciously break the compatibility with netstandard2.0 here and now, or are you saying we should leave behind the old, ineffective code as a fallback?

We can not break "compatibility with netstandard2.0" because the method is only for performance.
Yes, I ask you do not see back because it makes no sense - we have not real issues and real scenario to address and measure - measure it mandatory for performance changes. False positives on the old way will be reduced over time by themselves.

Reply:
That's not a bad point to make, I suppose. I don't really agree -- that method of checking is one of the fallback ones, so it's not going to be used in the primary case and thus is of secondary importance at best. I guess I don't see a compelling reason _not_ to check for that as well.

Reply:
>  I don't see a compelling reason

Do you already understand what you have done a breaking change and probably huge breaking change?
Interestingly, now that you have caught a bear, what will you do with it? :-)

Reply:
@iSazonov 

> Do you already understand what you have done a breaking change and probably huge breaking change?
> Interestingly, now that you have caught a bear, what will you do with it? :-)

I don't see any breaking changes here.  Can you provide an example?

Reply:
CIs fail on this. To save your time - it is Datetime.

Reply:
What do you mean? `[datetime]` declares the correct attribute and will be registered as not mutable by this code correctly. Are you saying it's somehow mutable?

Reply:
Yes.

Reply:
You're gonna have to elaborate on that one and why it affects this, I'm afraid. I don't follow. Datetime isn't mutable as far as I'm aware.

Reply:
What does the method return for the type before and after the change?

Reply:
If the method was returning `true` in this case before this change, then the method is actually wrong. That's part of why I made this PR, because the assumptions this method was using are very flawed. We can pretty easily verify that datetime is immutable:

![image](https://user-images.githubusercontent.com/32407840/80826382-1c171080-8bb0-11ea-8928-a7730bc11cc1.png)

None of its fields can be modified; it's not mutable. Any methods on the type simply return a new instance, and don't modify the original. The fact that it's decorated with the appropriate compiler attribute also confirms this.

Reply:
If we have code that is relying on these false assumptions somehow, I'm happy to fix that too. I intend to investigate the CI failures tonight.

Reply:
> CIs fail on this. To save your time - it is Datetime.

The only thing that is going to change with `datetime` is that when you assign one to a variable it will be done faster.

The current behavior is that it will be detected as mutable and a pointless copy will be done to ensure changes aren't persisted.  But it's immutable, there will be no changes, so no copy is required.

Reply:
@vexx32 Please revert the questionable change. The PR will be merged for non-LTS version and will affected most of users only after next PowerShell LTS - in the time this will be not actual I believe.